File: LCP 30.cpp
class Solution {
public:
    int magicTower(vector<int>& nums) {
        long long sum = 1;
        priority_queue<int, vector<int>, greater<int>> pq;
        int ans = 0;
        if (accumulate(nums.begin(), nums.end(), (long long)0) < 0) return -1;
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] < 0) {
                pq.push(nums[i]);
            }
            if (sum + nums[i] <= 0) {
                auto t = pq.top();
                pq.pop();
                sum += nums[i] - t;
                nums.push_back(t);
                ans++;
            } else {
                sum += nums[i];
            }
        }
        return ans;
    }
};

File: LCP 30.go
func magicTower(nums []int) int {
    h := hp{}
    heap.Init(&h)
    sum, ans := 1, 0
    for _, x := range nums {
        sum += x
    }
    if sum <= 0 {
        return -1
    } 
    sum = 1
    for _, x := range nums {
        if x < 0 {
            heap.Push(&h, x)
        }
        sum += x
        if sum <= 0 {
            sum -= heap.Pop(&h).(int)
            ans++
        }
    }

    return ans
}

type hp []int
func (h hp) Len() int {return len(h)}
func (h hp) Less(i, j int) bool {return h[i] < h[j]}
func (h hp) Swap(i, j int) {h[i], h[j] = h[j], h[i]}
func (h *hp) Push(x any) {
    *h = append(*h, x.(int))
}
func (h *hp) Pop() any {
    old := *h
    n := len(old)
    x := old[n-1]
    *h = old[:n-1]
    return x
}